packages/hardhat-core/src/internal/hardhat-network/provider/fork/ForkStateManager.ts

```diff
@@ -319,6 +329,10 @@ export class ForkStateManager implements PStateManager {
     if (this._stateCheckpoints.length !== 0) {
       throw checkpointedError("setBlockContext");
     }
+    if (blockNumber.eq(this._forkBlockNumber)) {
+      this._setStateRoot(toBuffer(this._initialStateRoot));
+      return;
+    }
```

Comment:
Instead of adding this if statement, we could make the `this._initialStateRoot` equal to the actual `stateRoot` of the fork block. However this would require fetching the block before FSM construction and would slow down the HardhatNode creation. Having considered this I would leave this the way it is.

Reply:
I did exactly the same reasoning, I'm glad that you agree :sweat_smile: